### ISSUE TITLE
docker build --no-cache doesn't update the cache - maybe should?

### ISSUE THREAD

**Issue by anentropic** [docker/docker#9880]
I have a python-based app which does `pip install -r requirements.txt` and one of the packages has been updated (I'm sure there are lots of other cases where my issue would apply too).

Since the Dockerfile hasn't changed I have to do `docker build --no-cache` to actually get the updated requirement into my image.

That all works fine.  Having built the image once with `--no-cache` to get the updated requirement I would expect that subsequent builds would use this as the base.

But that's not the case... if I now do a `docker build` it reverts to using the old version of the requirement.  It seems that `--no-cache` flag means not only that Docker doesn't use the cached fs layers, but it _doesn't update them_ either.

This is a bit of a nuisance, because I'm now condemned to doing `--no-cache` builds (with lengthy download and compile steps etc) from now on.

I expect there is some workaround to 'clear the cache' etc but it seems to me that `--no-cache` builds should update the cache rather than bypassing it entirely.

**Comment by duglin** [comment]
Are you sure its not working?
I just ran a quick test.
Dockerfile:
FROM ubuntu
RUN ls -l /tmp

And then when I run "docker build" twice, once w/o cache, and once with it, the 2nd one uses the cache created by the first:

$ docker build --no-cache .
Sending build context to Docker daemon 65.54 kB
Sending build context to Docker daemon 
Step 0 : FROM ubuntu
 ---> 86ce37374f40
Step 1 : RUN ls -l /tmp
 ---> Running in 5a455dffb386
total 0
 ---> 691b92772c06
Removing intermediate container 5a455dffb386
Successfully built 691b92772c06

Perhaps I need a different testcase to show the issue?
$ docker build .
Sending build context to Docker daemon 65.54 kB
Sending build context to Docker daemon 
Step 0 : FROM ubuntu
 ---> 86ce37374f40
Step 1 : RUN ls -l /tmp
 ---> Using cache
 ---> 691b92772c06
Successfully built 691b92772c06

Notice 691b92772c06 was created in the first build but used in the 2nd.

**Comment by anentropic** [comment]
what happens if you:

1. `docker build`
2. change content
3. `docker build --no-cache`
4. `docker build`

?

**Comment by duglin** [comment]
still seems to work.  The "change content" was to add a / to the end of /tmp .
In this case the build in step 4 used the output from the build in step 3.
I wonder if its just certain dockerfile commands.  Do you have a minimal Dockerfile that shows the issue?

**Comment by anentropic** [comment]
but in that case you've changed the Dockerfile, not the filesystem content, so the cache is invalidated anyway

I'll try and work up a minimal test case

**Comment by anentropic** [comment]
not sure it's possible a Dockerfile-only minimal test case

my issue is about a python package that has changed since step 1... my Dockerfile stays the same throughout but I need to regenerate the filesystem layers

same problem could apply to deb packages with `apt-get install` etc as well

**Comment by duglin** [comment]
Ah ha yes I see it now using ADD instead of RUN.   Still digging....

**Comment by duglin** [comment]
It looks like it is Dockerfile command specific.  ADD and COPY have logic to not even generate a hash if --no-cache is set, so it doesn't have the result in the cache.  From my testing if I force it to still generate a hash then things work.
@erikh what do you think about this?  Seems to me that we should still generate the cache entries, just skip looking for a previous one to use when --no-cache is specified.

**Comment by erikh** [comment]
SGTM; I think that's the desired behavior. Can you carry this?

**Comment by duglin** [comment]
Yup will do

-Doug

**Comment by anentropic** [comment]
can someone please reopen this, I don't think the fix in #9973 addressed my case:

Docker version 1.4.1, build 5bc2ff8

```
docker build -t myimage .
Step 9 : RUN /home/docker/venv/bin/pip install -r /home/docker/requirements.txt
 ---> Using cache
 ---> bb230588fbb7

docker build -t myimage --no-cache .
 Step 9 : RUN /home/docker/venv/bin/pip install -r /home/docker/requirements.txt
 ---> Running in 35bbdc155e64
 <output from pip>
 Cleaning up...
 ---> 4410e5630c81

docker build -t myimage .
Step 9 : RUN /home/docker/venv/bin/pip install -r /home/docker/requirements.txt
 ---> Using cache
 ---> bb230588fbb7
```

i.e. the third build has reverted to the first version of step 9, instead of the updated one that was generated by the `--no-cache` build

**Comment by duglin** [comment]
Please post the entire build output - I'd like to see what the steps before the RUN in step 9 are doing w.r.t. the cache.

**Comment by anentropic** [comment]
There's some `ONBUILD` triggers at the start from parent Dockerfile

I don't change the content of requirements.txt between runs

Step 8 is the workaround I've been using but otherwise irrelevant

1.
```
$ docker build -t anentropic/myapp:latest .
Sending build context to Docker daemon 142.3 MB
Sending build context to Docker daemon
Step 0 : FROM anentropic/uwsgi_app
# Executing 3 build triggers
Trigger 0, COPY anentropic/uwsgi_app.ini /etc/uwsgi/app_base.ini
Step 0 : COPY anentropic/uwsgi_app.ini /etc/uwsgi/app_base.ini
 ---> Using cache
Trigger 1, COPY anentropic/*.sh /home/docker/scripts/
Step 0 : COPY anentropic/*.sh /home/docker/scripts/
 ---> Using cache
Trigger 2, RUN chmod -R +x /home/docker/scripts
Step 0 : RUN chmod -R +x /home/docker/scripts
 ---> Using cache
 ---> c1ad2a4f0ab4
Step 1 : MAINTAINER anentropic
 ---> Using cache
 ---> 502612c1b9c7
Step 2 : RUN apt-get update
 ---> Using cache
 ---> 6fab0e9aa941
Step 3 : RUN apt-get install -y --fix-missing libpq-dev
 ---> Using cache
 ---> acf6c54ca909
Step 4 : RUN apt-get install -y libjpeg-dev
 ---> Using cache
 ---> 22475efc14db
Step 5 : RUN apt-get install -y libxml2-dev libxslt-dev zlib1g-dev
 ---> Using cache
 ---> 75f5e1f474d0
Step 6 : RUN apt-get install -y binutils libproj-dev gdal-bin
 ---> Using cache
 ---> 1742c09474cf
Step 7 : RUN apt-get install -y spatialite-bin
 ---> Using cache
 ---> e1e5b345928d
Step 8 : RUN echo "cache bust: 2"
 ---> Using cache
 ---> 10d533979d42
Step 9 : COPY ./requirements.txt /home/docker/requirements.txt
 ---> Using cache
 ---> 76b640d7adf9
Step 10 : RUN /home/docker/venv/bin/pip install -r /home/docker/requirements.txt
 ---> Using cache
 ---> 7ce6a7c1209d
Step 11 : RUN rm /home/docker/requirements.txt
 ---> Using cache
 ---> 94693abc1c2e
Step 12 : COPY ./ /home/docker/app
 ---> a70b83a84c2d
Removing intermediate container f573d40745da
Step 13 : RUN find /home/docker/app -name '*.pyc' -delete
 ---> Running in 741cfd0a9840
 ---> 1e1d0cdd4ef4
Removing intermediate container 741cfd0a9840
Step 14 : RUN find /home/docker/app -maxdepth 1 -name "_*" -print0 | xargs -0 rm -rf
 ---> Running in 613b368a5523
 ---> 84e655aee996
Removing intermediate container 613b368a5523
Step 15 : RUN find /home/docker/app -maxdepth 1 -name ".*" -print0 | xargs -0 rm -rf
 ---> Running in 9fea60831ac5
 ---> a93b1f529cc9
Removing intermediate container 9fea60831ac5
Step 16 : RUN ln -s /home/docker/app/manage_docker.py /home/docker/app/manage.py
 ---> Running in 1088cd3d277f
 ---> 756688066cb6
Removing intermediate container 1088cd3d277f
Step 17 : RUN mkdir /home/docker/media
 ---> Running in 13726065bb43
 ---> acdf70693644
Removing intermediate container 13726065bb43
Step 18 : RUN mkdir /home/docker/static
 ---> Running in 5bc1d61e4a74
 ---> b0e193277d82
Removing intermediate container 5bc1d61e4a74
Step 19 : ENTRYPOINT /home/docker/scripts/start_uwsgi_app.sh --ini /home/docker/app/anentropic/uwsgi_app.ini
 ---> Running in 30111f47eb9c
 ---> f372e9babaea
Removing intermediate container 30111f47eb9c
Successfully built f372e9babaea
```

note step 10 `---> 7ce6a7c1209d`

2.
```
$ docker build -t anentropic/myapp:latest --no-cache .
Sending build context to Docker daemon 142.3 MB
Sending build context to Docker daemon
Step 0 : FROM anentropic/uwsgi_app
# Executing 3 build triggers
Trigger 0, COPY anentropic/uwsgi_app.ini /etc/uwsgi/app_base.ini
Step 0 : COPY anentropic/uwsgi_app.ini /etc/uwsgi/app_base.ini
Trigger 1, COPY anentropic/*.sh /home/docker/scripts/
Step 0 : COPY anentropic/*.sh /home/docker/scripts/
Trigger 2, RUN chmod -R +x /home/docker/scripts
Step 0 : RUN chmod -R +x /home/docker/scripts
 ---> Running in d7a43fc2e051
 ---> 943e15f29537
Removing intermediate container 1214a99f7761
Removing intermediate container d7a43fc2e051
Removing intermediate container e0a50aceceba
Step 1 : MAINTAINER anentropic
 ---> Running in 991ae8d6d777
 ---> 4ab6f02fa45f
Removing intermediate container 991ae8d6d777
Step 2 : RUN apt-get update
 ---> Running in 774b55404b26
 ...
 ---> e2e1c5de02cb
Removing intermediate container 774b55404b26
Step 3 : RUN apt-get install -y --fix-missing libpq-dev
 ---> Running in 1edecd8d6d1f
 ...
 ---> 93161336b93e
Removing intermediate container 1edecd8d6d1f
Step 4 : RUN apt-get install -y libjpeg-dev
 ---> Running in 6d1cfd27a558
 ...
 ---> 4d000d13d8fb
Removing intermediate container 6d1cfd27a558
Step 5 : RUN apt-get install -y libxml2-dev libxslt-dev zlib1g-dev
 ---> Running in 3f693e9987b8
 ...
 ---> 342ba6fcb2bf
Removing intermediate container 3f693e9987b8
Step 6 : RUN apt-get install -y binutils libproj-dev gdal-bin
 ---> Running in 21722bb922dc
 ...
 ---> c76f05bcc512
Removing intermediate container 21722bb922dc
Step 7 : RUN apt-get install -y spatialite-bin
 ---> Running in 02d1fe245717
 ...
 ---> f154482b1a22
Removing intermediate container 02d1fe245717
Step 8 : RUN echo "cache bust: 2"
 ---> Running in 76e007739268
cache bust: 2
 ---> 0e7afa5f31e1
Removing intermediate container 76e007739268
Step 9 : COPY ./requirements.txt /home/docker/requirements.txt
 ---> b9c145b0f01e
Removing intermediate container c0fe34914a91
Step 10 : RUN /home/docker/venv/bin/pip install -r /home/docker/requirements.txt
 ---> Running in 390eb19426ff
 ...
 ---> 8def99ba58b1
Removing intermediate container 390eb19426ff
Step 11 : RUN rm /home/docker/requirements.txt
 ---> Running in 38b36b3b4e5c
 ---> 071e9642a0d3
Removing intermediate container 38b36b3b4e5c
Step 12 : COPY ./ /home/docker/app
 ---> 555e8f915bcc
Removing intermediate container 7a1379282095
Step 13 : RUN find /home/docker/app -name '*.pyc' -delete
 ---> Running in 228d45b9c53b
 ---> dd45ad2c9233
Removing intermediate container 228d45b9c53b
Step 14 : RUN find /home/docker/app -maxdepth 1 -name "_*" -print0 | xargs -0 rm -rf
 ---> Running in ccd3d7b20a64
 ---> e639772fde3b
Removing intermediate container ccd3d7b20a64
Step 15 : RUN find /home/docker/app -maxdepth 1 -name ".*" -print0 | xargs -0 rm -rf
 ---> Running in e9645de930b1
 ---> 1dc9e0809585
Removing intermediate container e9645de930b1
Step 16 : RUN ln -s /home/docker/app/manage_docker.py /home/docker/app/manage.py
 ---> Running in b43d5d63e582
 ---> 9322678e41f4
Removing intermediate container b43d5d63e582
Step 17 : RUN mkdir /home/docker/media
 ---> Running in 591fb1c03170
 ---> 73efcd237795
Removing intermediate container 591fb1c03170
Step 18 : RUN mkdir /home/docker/static
 ---> Running in 61113ccc8ac2
 ---> c7b4b1921986
Removing intermediate container 61113ccc8ac2
Step 19 : ENTRYPOINT /home/docker/scripts/start_uwsgi_app.sh --ini /home/docker/app/anentropic/uwsgi_app.ini
 ---> Running in 2e84b60d6023
 ---> d516bc430a11
Removing intermediate container 2e84b60d6023
Successfully built d516bc430a11
```

note step 10 ` ---> 8def99ba58b1`

3.
```
$ docker build -t anentropic/myapp:latest .
Sending build context to Docker daemon 142.3 MB
Sending build context to Docker daemon
Step 0 : FROM anentropic/uwsgi_app
# Executing 3 build triggers
Trigger 0, COPY anentropic/uwsgi_app.ini /etc/uwsgi/app_base.ini
Step 0 : COPY anentropic/uwsgi_app.ini /etc/uwsgi/app_base.ini
 ---> Using cache
Trigger 1, COPY anentropic/*.sh /home/docker/scripts/
Step 0 : COPY anentropic/*.sh /home/docker/scripts/
 ---> Using cache
Trigger 2, RUN chmod -R +x /home/docker/scripts
Step 0 : RUN chmod -R +x /home/docker/scripts
 ---> Using cache
 ---> c1ad2a4f0ab4
Step 1 : MAINTAINER anentropic
 ---> Using cache
 ---> 502612c1b9c7
Step 2 : RUN apt-get update
 ---> Using cache
 ---> 6fab0e9aa941
Step 3 : RUN apt-get install -y --fix-missing libpq-dev
 ---> Using cache
 ---> acf6c54ca909
Step 4 : RUN apt-get install -y libjpeg-dev
 ---> Using cache
 ---> 22475efc14db
Step 5 : RUN apt-get install -y libxml2-dev libxslt-dev zlib1g-dev
 ---> Using cache
 ---> 75f5e1f474d0
Step 6 : RUN apt-get install -y binutils libproj-dev gdal-bin
 ---> Using cache
 ---> 1742c09474cf
Step 7 : RUN apt-get install -y spatialite-bin
 ---> Using cache
 ---> e1e5b345928d
Step 8 : RUN echo "cache bust: 2"
 ---> Using cache
 ---> 10d533979d42
Step 9 : COPY ./requirements.txt /home/docker/requirements.txt
 ---> Using cache
 ---> 76b640d7adf9
Step 10 : RUN /home/docker/venv/bin/pip install -r /home/docker/requirements.txt
 ---> Using cache
 ---> 7ce6a7c1209d
Step 11 : RUN rm /home/docker/requirements.txt
 ---> Using cache
 ---> 94693abc1c2e
Step 12 : COPY ./ /home/docker/app
 ---> b2a2d3b7255e
Removing intermediate container e7adb5d2c831
Step 13 : RUN find /home/docker/app -name '*.pyc' -delete
 ---> Running in 6c5399be258e
 ---> d89fe9cfa1b4
Removing intermediate container 6c5399be258e
Step 14 : RUN find /home/docker/app -maxdepth 1 -name "_*" -print0 | xargs -0 rm -rf
 ---> Running in 6a2943b546b8
 ---> be074a3e2451
Removing intermediate container 6a2943b546b8
Step 15 : RUN find /home/docker/app -maxdepth 1 -name ".*" -print0 | xargs -0 rm -rf
 ---> Running in 11df3ba9c225
 ---> 5ea7713ddc99
Removing intermediate container 11df3ba9c225
Step 16 : RUN ln -s /home/docker/app/manage_docker.py /home/docker/app/manage.py
 ---> Running in 9187ab77893b
 ---> cd9c567bded1
Removing intermediate container 9187ab77893b
Step 17 : RUN mkdir /home/docker/media
 ---> Running in 8a06d1f7cb53
 ---> 19d978ffd50c
Removing intermediate container 8a06d1f7cb53
Step 18 : RUN mkdir /home/docker/static
 ---> Running in 993469b4942d
 ---> afcd21b8f6f3
Removing intermediate container 993469b4942d
Step 19 : ENTRYPOINT /home/docker/scripts/start_uwsgi_app.sh --ini /home/docker/app/anentropic/uwsgi_app.ini
 ---> Running in 65574378dd26
 ---> 84d8fb566265
Removing intermediate container 65574378dd26
Successfully built 84d8fb566265
```

note step 10 ` ---> 7ce6a7c1209d` ...same as 1st build

**Comment by duglin** [comment]
I think I've been able to reproduce this and its related to the ONBUILD processing, I think.
Here's the setup:
file foo:
Testing

file base:
FROM ubuntu
ONBUILD COPY foo /tmp/

file: Dockerfile
FROM base
RUN pwd

$ docker build -f base -t base .
$ docker build .
$ docker build --no-cache .
$ docker build .

Should show that the cache from the "no-cache" build is not used during the final build.
When I combine it all into one Dockerfile:
FROM ubuntu
COPY foo /tmp/
RUN pwd

it works as expected.  Now on to the debugging....   will reopen issue.

**Comment by duglin** [comment]
@anentropic please let me know if you think this doesn't match your scenario.
Also, note the use of the "-f" flag  :-)
And, when I replace COPY with RUN in the 'base' dockerfile it works, so I think its related to the COPY cache processing.

**Comment by duglin** [comment]
ok I may have jumped the gun (not the shark) here.  I think I reproduced this while running with an old docker daemon.  While debugging I started to use the latest docker daemon and now I can't reproduce it.  The difference is that the old docker didn't include the COPY fix from #9973 .  So, now I'm thinking that #9973 really did fix it.  @anentropic have you tested out the fix to see that it doesn't fix your issue?  It looks like the output you posted was from a Dec 16th version - which didn't include my fix.

**Comment by anentropic** [comment]
this looks like it addresses my issue

yes I am still using released version 1.4.1

I am on OS X so am dependent on boot2docker

**Comment by duglin** [comment]
@anentropic can you try with my fix? I'm pretty sure it fixes your issue.

**Comment by thaJeztah** [comment]
^^ @anentropic if you want to try the current "master" version of docker, pre-built binaries can be found at https//master.dockerproject.com.

**Comment by duglin** [comment]
@anentropic is this fixed?

**Comment by anentropic** [comment]
@thaJeztah ...TBH I'd need instructions for how to install the pre-built binary into my boot2docker vm

**Comment by thaJeztah** [comment]
Ah, right, not a boot2docker user myself. Tempted to say "ssh into boot2docker, replace the binary and restart the daemon", but not sure that'd work :smile:

**Comment by duglin** [comment]
I'm pretty sure that would work

**Comment by anentropic** [comment]
Ah ok it's easier than I thought. So, my starting point:

```
Boot2Docker version 1.4.1, build master : 86f7ec8 - Tue Dec 16 23:11:29 UTC 2014
Docker version 1.4.1, build 5bc2ff8
docker@boot2docker:~$ which docker
/usr/local/bin/docker
```

I downloaded `linux/amd64/docker-1.4.1-dev` and replaced `/usr/local/bin/docker` with it, I now have:

```
$ docker -v
Docker version 1.4.1-dev, build 27602f2
```

However this wasn't enough, running the builds again I noticed no difference.  It's because the docker daemon was already running.  I had to do a `sudo /etc/init.d/docker restart`

__NOTE:__  
for anyone else trying the same _do not restart your boot2docker vm_ (i.e. don't `boot2docker restart` from the OS X host) as this seems to restore the original filesystem, i.e. it wipes out the newly downloaded version of docker. Just do an init.d restart as above

Anyway, happy to report that the caching issue is fixed with current dev version of docker
